README.md
# libpca9685

C driver for the NXP Semiconductors PCA9685 16-channel 12-bit PWM I2C-bus LED Controller ([datasheet](https://github.com/carlodicelico/libpca9685/tree/master/docs/PCA9685.pdf))

## Features

- Tests with [greatest](https://github.com/silentbicycle/greatest) and [theft](https://github.com/silentbicycle/theft)
- Bring your own I2C and hardware interface libraries
- Minimal dependencies

## Goals

- Strict, full implementation of manufacturer specifications
- A clean, user-friendly interface for application developers and library writers
- No tight coupling to a specific I2C library or hardware platform (Linux I2C, various Pi/Arduino libraries, etc.)
- No tight coupling to a specific application (LEDs vs. servos, etc.)
- Modern C (C11) development practices, like unit tests, property-based tests, [modern CMake (3.13)](https://cliutils.gitlab.io/modern-cmake/), etc.

## Installation

**Option 1:** add it as a git submodule like this:

```shell
$ git clone git@github.com:carlodicelico/libpca9685.git vendor/libpca9685
```

Then, add it in your `CMakeLists.txt` and link your library or executable:
```cmake
add_subdirectory(vendor/libpca9685)
target_link_libraries(my_lib pca9685)
```

**Option 2:** install it like this:

```shell
$ git clone git@github.com:carlodicelico/libpca9685.git && cd libpca9685
$ cmake -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -G "Unix Makefiles" -S . -B cmake-build-release
$ cd cmake-build-release
$ make && sudo make install 
```

Then, add it in your `CMakeLists.txt` and link your library or executable:

```cmake
find_library(PCA9506 pca9685)
target_link_libraries(my_lib ${PCA9685})
```

## Usage

Example application code might look something like this:

```C
#include "libpca9685.h"

// Application (not library) dependencies
#include "pigpio.h"
#include <zconf.h>

#define I2C_BUS 1
#define I2C_BUS_ADDRESS 0x40
#define I2C_FLAGS 0

// ALL management of the I2C bus itself is your application's responsibility!
static int pi, handle;

// Set up an I2C reader callback
static uint8_t i2c_read_byte_cb(pca9685_s *driver, uint8_t r) {
    // This is pigpio's function we're wrapping here, but it could be any other library, as well
    int result = i2c_read_byte_data(pi, (uint8_t)servo_master_handle, r);

    // You can add information to the driver struct's status field if you like
    driver->status = (result >= 0) ? "ok" : pigpio_error(result);

    return (uint8_t)result;
}

static uint8_t i2c_write_byte_cb(pca9685_s *driver, uint8_t r, uint8_t d) {
    // This is pigpio's function we're wrapping here, but it could be any other library, as well
    int result = i2c_write_byte_data(pi, (uint8_t)servo_master_handle, r, d);

    // You can add information to the driver struct's status field if you like
    driver->status = (result >= 0) ? "ok" : pigpio_error(result);

    return (uint8_t)result;
}

int main(int argc, char **argv) {
    pi = pigpio_start(NULL, NULL);
    char *pi_status = (pi >= 0) ? "ok" : pigpio_error(pi);

    handle = i2c_open(pi, I2C_BUS, I2C_BUS_ADDRESS, I2C_FLAGS);
    char *handle_status = (handle >= 0) ? "ok" : pigpio_error(handle);

    printf("\nMain\n====\n: %d");
    printf("\nPIGPIO Status: %s", pi_status);
    printf("\nI2C Handle Status: %s\n", handle_status);
    printf("Initializing driver...");

    // Initialize your driver struct—also does test read and write operations
    pca9685_s my_driver = pca9685(.bus_reader=bus_byte_reader, .bus_writer=bus_byte_writer);

    printf("%s\n", my_driver.status);

    // Set the chip's frequency to 300Hz
    my_driver.set_frequency(&my_driver, 300);

    // Set up a basic timer for example purposes (10 minutes, in seconds)
    int simple_timer = 600;

    // Flash all LEDs on and off continuously until the timer runs out
    while(simple_timer > 0) {
        my_driver.channel_on(&my_driver, -1);
        sleep(1);
        my_driver.channel_off(&my_driver, -1);
        sleep(1);

        simple_timer--;
    }

    // Do some other stuff here...

    // You can attempt an orderly shutdown
    my_driver.soft_reset(&my_driver);

    // Or maybe just restore all defaults
    my_driver.hard_reset(&my_driver);

    // pigpio-specific...
    int close = i2c_close(pi, (uint)handle);
    char *close_message = (close == PI_BAD_HANDLE) ? "PI_BAD_HANDLE" : "CLOSED";
    printf("\nClose result: %s\n", close_message);

    pigpio_stop(pi);

    if(close == PI_BAD_HANDLE) return 1;

    return 0;
}
```

See also the included [examples](https://github.com/carlodicelico/libpca9685/tree/master/examples).

## How do I contribute?

Contributions are welcome! Check out [CONTRIBUTING](https://github.com/carlodicelico/libpca9685/tree/master/CONTRIBUTING.md) for instructions.

## How is this library versioned?

This library follows the principles of [Semantic Versioning](http://semver.org/). You can find each new release,
along with the changelog, in the [CHANGELOG](https://github.com/carlodicelico/libpca9685/tree/master/CHANGELOG.md).

During initial development, the major version will be 0 (e.g., `0.x.y`), which indicates the code does not yet have a
stable API. Once we hit `1.0.0`, we will make every effort to maintain a backwards compatible API and use the MAJOR,
MINOR, and PATCH versions on each release to indicate any incompatibilities.

## License

This code is released under the MIT License. Please see [LICENSE](https://github.com/carlodicelico/libpca9685/tree/master/LICENSE) for more details.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- TBD

## v0.2.2
- Update documentation, license
## v0.2.1
- Make installable
## v0.2.0
### Added
- Tests for calculations
### Changed
- Moved driver functions into struct for a cleaner API and namespacing
- Improved tests

## v0.1.0
### Added
- Project setup, including:
  - CMake setup
  - Unit and property-based tests
- Driver library, including:
  - Interface for user-provided I2C read/write callbacks
  - Basic data structure for the driver
  - Definitions for key manufacturer register addresses and default values
  - Calculations, including:
    - Calculate a register address from a provided channel (0-based)
    - Calculate a prescale value from a provided frequency
    - Bitwise operations for changing LED on and off values
  - Write to individual registers and all registers, including:
    - Set chip base PWM frequency
    - Set duty cycle per channel
    - Set duty cycle on all channels
    - Reset driver
    - Wait the correct amount of time (at least 500μs) after enabling the oscillator
  - Public API, including:
    - set frequency
    - set duty cycle
    - channel on/off
    - soft/hard resets
  - Non-destructive (soft) reset
  - Destructive (hard) reset
- Documentation, including:
  - this CHANGELOG
  - [AUTHORS](https://github.com/carlodicelico/libpca9685/tree/master/AUTHORS)
  - [CONDUCT](https://github.com/carlodicelico/libpca9685/tree/master/CONDUCT.md)
  - [CONTRIBUTING](https://github.com/carlodicelico/libpca9685/tree/master/CONTRIBUTING.md)
  - [LICENSE](https://github.com/carlodicelico/libpca9685/tree/master/LICENSE)
  - [README](https://github.com/carlodicelico/libpca9685/tree/master/README.md)


tests/runner.c
#include "runner.h"

GREATEST_MAIN_DEFS();

int main(int argc, char **argv) {
	(void)argc;
	(void)argv;

	struct greatest_report_t report;

	GREATEST_INIT();
	greatest_set_verbosity(3);

	RUN_SUITE(test_driver_init);
    RUN_SUITE(test_register_ops);

	GREATEST_PRINT_REPORT();

	greatest_get_report(&report);

	if(report.failed > 0) return 1;

	return 0;
}


tests/suites/tests.c
#include "tests.h"

#include <stdint.h>

mock_register_s mock_driver;

void set_up_mock_driver(pca9685_s *driver) {
    *driver = pca9685(.bus_reader=mock_bus_reader, .bus_writer=mock_bus_writer);
}

void set_up_theft_run_config_u8(struct theft_run_config *config) {
    theft_seed seed = theft_seed_of_time();

    *config = (struct theft_run_config){
        .type_info = {
            theft_get_builtin_type_info(THEFT_BUILTIN_uint8_t)
        },
        .seed = seed,
        .trials = 100
    };
}

void set_up_theft_run_config_int(struct theft_run_config *config) {
    theft_seed seed = theft_seed_of_time();

    *config = (struct theft_run_config){
        .type_info = {
            theft_get_builtin_type_info(THEFT_BUILTIN_int)
        },
        .seed = seed,
        .trials = 100
    };
}

void set_up_theft_run_config_int_int(struct theft_run_config *config) {
    theft_seed seed = theft_seed_of_time();

    *config = (struct theft_run_config){
        .type_info = {
            theft_get_builtin_type_info(THEFT_BUILTIN_int),
            theft_get_builtin_type_info(THEFT_BUILTIN_int)
        },
        .seed = seed,
        .trials = 100
    };
}

void set_up_theft_run_config_u8_int_int(struct theft_run_config *config) {
    theft_seed seed = theft_seed_of_time();

    *config = (struct theft_run_config){
        .type_info = {
            theft_get_builtin_type_info(THEFT_BUILTIN_uint8_t),
            theft_get_builtin_type_info(THEFT_BUILTIN_int),
            theft_get_builtin_type_info(THEFT_BUILTIN_int)
        },
        .seed = seed,
        .trials = 100
    };
}

u8 mock_bus_reader(pca9685_s *driver, u8 address) {
    (void)driver;

    mock_driver.address = address;
    return mock_driver.value;
}

u8 mock_bus_writer(pca9685_s *driver, u8 address, u8 data_in) {
    (void)driver;

    mock_driver.address = address;
    mock_driver.value = data_in;

    return 0;
}



tests/suites/tests.h
#ifndef TESTS_H
#define TESTS_H

#include "greatest.h"
#include "theft.h"

#include "pca9685.h"
#include "registers.h"

typedef struct mock_register {
    u8 address;
    u8 value;
} mock_register_s;

void set_up_mock_driver(pca9685_s *);
void set_up_theft_run_config_u8(struct theft_run_config *config);
void set_up_theft_run_config_int(struct theft_run_config *config);
void set_up_theft_run_config_int_int(struct theft_run_config *config);
void set_up_theft_run_config_u8_int_int(struct theft_run_config *config);
u8 mock_bus_reader(pca9685_s *, u8 address);
u8 mock_bus_writer(pca9685_s *, u8 address, u8 data_in);

#endif


src/pca9685.c
#include "pca9685.h"
#include "registers.h"

#include <stdio.h>

static void set_frequency(pca9685_s *h, int frequency) {
    set_pwm_frequency(h, frequency);
}

static void set_duty_cycle(pca9685_s *h, int channel, int delay, int percent) {
   set_pwm_duty_cycle(h, channel, delay, percent);
}

static void channel_on(pca9685_s *h, int channel) {
    set_pwm_duty_cycle(h, channel, 0, 100);
}

static void channel_off(pca9685_s *h, int channel) {
    set_pwm_duty_cycle(h, channel, 0, 0);
}

static void soft_reset(pca9685_s *h) {
    reset_driver_soft(h);
}

static void hard_reset(pca9685_s *h) {
    reset_driver_hard(h);
}

pca9685_s pca9685_configure_handle(pca9685_s handle){
    if(!(handle.bus_reader && handle.bus_writer)){
        handle.command = "cb_check";
        handle.status = "Invalid input: bus reader and bus writer callbacks are both required.";
    } else {
        handle.command = "rw_check";

        // Quick & dirty read/write checks; only verifies the interface, should break on the caller's side.
        pca9685_i2c_bus_read(&handle, LED0_ON_L);
        pca9685_i2c_bus_write(&handle, LED0_ON_L, handle.data);
    }

    handle.soft_reset = soft_reset;
    handle.hard_reset = hard_reset;
    handle.set_frequency = set_frequency;
    handle.channel_on = channel_on;
    handle.channel_off = channel_off;
    handle.set_duty_cycle = set_duty_cycle;

    handle.command = handle.command == NULL ? "init" : handle.command;
    handle.status = handle.status == NULL ? "ok" : handle.status;

    return handle;
}



src/libpca9685.c
#include "libpca9685.h"
#include "registers.h"

#include <stdio.h>

static void set_frequency(pca9685_s *h, int frequency) {
    set_pwm_frequency(h, frequency);
}

static void set_duty_cycle(pca9685_s *h, int channel, int delay, int percent) {
   set_pwm_duty_cycle(h, channel, delay, percent);
}

static void channel_on(pca9685_s *h, int channel) {
    set_pwm_duty_cycle(h, channel, 1, 100);
}

static void channel_off(pca9685_s *h, int channel) {
    set_pwm_duty_cycle(h, channel, 1, 0);
}

static void soft_reset(pca9685_s *h) {
    reset_driver_soft(h);
}

static void hard_reset(pca9685_s *h) {
    reset_driver_hard(h);
}

pca9685_s pca9685_configure_handle(pca9685_s handle){
    if(!(handle.bus_reader && handle.bus_writer)){
        handle.command = "cb_check";
        handle.status = "Invalid input: bus reader and bus writer callbacks are both required.";
    } else {
        handle.command = "rw_check";

        // Quick & dirty read/write checks; only verifies the interface, should break on the caller's side.
        pca9685_i2c_bus_read(&handle, LED0_ON_L);
        pca9685_i2c_bus_write(&handle, LED0_ON_L, handle.data);
    }

    handle.soft_reset = soft_reset;
    handle.hard_reset = hard_reset;
    handle.set_frequency = set_frequency;
    handle.channel_on = channel_on;
    handle.channel_off = channel_off;
    handle.set_duty_cycle = set_duty_cycle;

    handle.command = handle.command == NULL ? "init" : handle.command;
    handle.status = handle.status == NULL ? "ok" : handle.status;

    return handle;
}



include/registers.h
#ifndef PCA9685_REGISTERS_H
#define PCA9685_REGISTERS_H

#include <inttypes.h>
#include "pca9685.h"

/** Conveniences */
#define ZERO 0x00
#define LOW  ZERO
#define OK   ZERO

#define ONE  0x01
#define HIGH ONE
#define ERR  ONE

/** For flags */
#define ALL -1

/** Bus logic bits: 0 for write/enabled, 1 for read/disabled */
#define BUS_ENABLED  ZERO
#define BUS_DISABLED ONE

/** LED bits: 0 for completely off, 1 for completely on */
#define LED_OFF LOW
#define LED_ON  HIGH

/** The PCA9685 has 12-bit resolution, or 4096 (0–4095) steps from full off to full on */
#define LED_MAX_BITS  4095
#define LED_MAX_STEPS (LED_MAX_BITS + 1)

/** Oscillator clock frequency is 25MHz */
#define OSCILLATOR 25000000.0

/** Osclillator takes 500μs (500000ns) to switch states */
#define BEAT 500000

/* To calculate the remaining LED register addresses for a single channel,
 * we only need to calculate the base and add 1–3. To calculate the base, LED_OFFSET + (MULTIPLIER * channel).
 *
 * \see registers.c */
#define LED_OFFSET         6
#define MAX_CHANNEL       15
#define MIN_CHANNEL        0
#define MULTIPLIER         4
#define PRESCALE_MIN       3
#define PRESCALE_MAX     255
#define FREQ_MIN          24
#define FREQ_MAX        1526

/** Register definitions - all defaults are in base16, and all flags are expressed as bit shifts */

/** Mode registers */
#define MODE1 ZERO
#define MODE2 ONE

/** I2C-bus subaddress registers */
#define SUBADR1 0x02
#define SUBADR2 0x03
#define SUBADR3 0x04

/* I2C-bus subaddress register defaults */
#define SUBADR1_DEFAULTS 0xe2
#define SUBADR2_DEFAULTS 0xe4
#define SUBADR3_DEFAULTS 0xe8

/* LED All Call I2C-bus address register */
#define ALLCALLADR 0x05

/* LED All Call I2C-bus address register defaults */
#define ALLCALLADR_DEFAULTS 0xe0

/* LED 0 output and brightness control registers */
#define LED0_ON_L 0x06

/** Registers for loading all LED registers by byte */
#define ALL_LED_ON_L  0xFA

/** Prescaler register for PWM output frequency */
#define PRE_SCALE 0xFE

/** Bits for Mode register 1: 8 bits total */
#define ALLCALL 1<<0
#define SUB3    1<<1
#define SUB2    1<<2
#define SUB1    1<<3
#define SLEEP   1<<4
#define AI      1<<5
#define EXTCLK  1<<6
#define RESTART 1<<7

/** Mode 1 register default and sleep disabled settings */
#define MODE1_DEFAULTS 0x88
#define MODE1_NO_SLEEP 0x01

/** Bits for mode register 2: 8 bits total (bits 5–7 are unused) */
#define OUTNE0 1<<0
#define OUTNE1 1<<1
#define OUTDRV 1<<2
#define OCH    1<<3
#define INVRT  1<<4

/** Mode 2 register defaults */
#define MODE2_DEFAULTS 0x40

/** Channel to register base conversion.
 * Takes a zero-based channel (like 10) and returns the first register (0x32). */
uint8_t channel_to_register_base(uint8_t channel);

/* Calculations */
int calculate_prescale_from_frequency(int frequency);
int calculate_delay_time_from_percentage(int delay);
int calculate_on_time_from_percentage(int percent);
int calculate_off_time_from_delay_and_on_time(int delay, int on_time);

/** Low-level access to user callback wrappers for initialization */
void pca9685_i2c_bus_read(pca9685_s *, uint8_t r);
void pca9685_i2c_bus_write(pca9685_s *, uint8_t r, uint8_t d);

/** Low-level access to register writes for testing */
void set_led_bytes(pca9685_s *, int c, int on, int off);

/** Reset driver state */
void reset_driver_soft(pca9685_s *h);
void reset_driver_hard(pca9685_s *h);

/** Sets the value of the PRE_SCALE register with the provided PWM output frequency */
void set_pwm_frequency(pca9685_s *h, int frequency);

/** Sets the PWM duty cycle with a % delay at a provided % on time */
void set_pwm_duty_cycle(pca9685_s *h, int channel, int delay, int percent);

#endif



include/libpca9685.h
/**
 * \file libpca9685.h
 *
 * \brief Driver for the NXP Semiconductors PCA9685 16-channel 12-bit PWM I2C-bus LED Controller
 *
 * \see https://www.nxp.com/docs/en/data-sheet/PCA9685.pdf
 * \see https://github.com/torvalds/linux/blob/master/include/linux/i2c.h#L152
 *
 * Usage:
 *
 * \code
 * // Set up your driver, passing your I2C read and write callbacks
 * my_driver = pca9685(.bus_reader=my_bus_reader, .bus_writer=my_bus_writer);
 *
 * // Set frequency to 250Hz
 * my_driver.set_frequency(&my_driver, 250);
 *
 * // Set PWM duty cycle on LED channel 5 to 25%
 * my_driver.set_duty_cycle(&my_driver, 5, 1, 25);
 *
 * // Turn LED channel on
 * my_driver.channel_on(&my_driver, 5);
 *
 * // Turn LED channel off
 * my_driver.channel_off(&my_driver, 5);
 *
 * // Turn all LEDs on
 * my_driver.channel_on(&my_driver, -1);
 *
 * // Turn all LEDs off
 * my_driver.channel_off(&my_driver, -1);
 *
 * // Set PWM duty cycle on all LED channels to 50%
 * my_driver.set_duty_cycle(&my_driver, -1, 1, 50);
 *
 * // Contains the data last read or written
 * uint8_t my_data = my_driver.data(&my_driver);
 *
 * // Contains the last register address read to or written from
 * uint8_t my_address = my_driver.address(&my_driver);
 *
 * // Contains the last command attempted
 * char *my_command = my_driver.command(&my_driver);
 *
 * // Contains the last command's status or an error message
 * char *my_status = my_driver.status(&my_driver);
 * \endcode
 */

#ifndef PCA9685_LIBRARY_H
#define PCA9685_LIBRARY_H

#include <inttypes.h>

/** Initializes a driver_handle struct with values provided by the user. */
#define pca9685(...) (pca9685_configure_handle((pca9685_s){__VA_ARGS__}))

/** Primitive data type aliases for readability & conciseness */
typedef char *string;
typedef uint8_t u8;

/** Aggregate types */
struct pca9685_driver;

/**
 * Types for the I2C bus read and write callbacks.
 * Modeled on the I2C interface for Linux (\c i2c_smbus_read_byte_data, etc.)
 *
 * \see https://github.com/torvalds/linux/blob/master/include/linux/i2c.h#L152
 */
typedef u8 (*pca9685_i2c_bus_read_cb)(struct pca9685_driver *driver, u8 address);
typedef u8 (*pca9685_i2c_bus_write_cb)(struct pca9685_driver *driver, u8 address, u8 data);

/** Public API function signatures */
typedef void (*pca9685_fn)(struct pca9685_driver *driver);
typedef void (*pca9685_chan_freq_fn)(struct pca9685_driver *driver, int chan_or_freq);
typedef void (*pca9685_duty_cycle_fn)(struct pca9685_driver *, int led_channel, int delay_percent, int percent_on);

/** Type definition for the driver handle */
typedef struct pca9685_driver {
    u8 data;                               // Data last read/written
    u8 address;                            // Last register read to/written from
    string command;                        // Last command attempted
    string status;                         // Status (ok) or error message
    pca9685_i2c_bus_read_cb bus_reader;    // I2C Bus reader callback
    pca9685_i2c_bus_write_cb bus_writer;   // I2C Bus writer callback
    pca9685_fn soft_reset;                 // Attempts an orderly shutdown, saving register values
    pca9685_fn hard_reset;                 // Resets to manufacturer defaults
    pca9685_chan_freq_fn set_frequency;    // Set output frequency; default is 200Hz
    pca9685_chan_freq_fn channel_on;       // Set LED channel on; ALL (-1) acts on all channels
    pca9685_chan_freq_fn channel_off;      // Set LED channel off; ALL (-1) acts on all channels
    pca9685_duty_cycle_fn set_duty_cycle;  // Set duty cycle % and delay % for more control; ALL (-1) acts on all channels
} pca9685_s;

/** Private, used by the macro defined above. */
pca9685_s pca9685_configure_handle(pca9685_s);

#endif


src/registers.c
#include "registers.h"

#include <inttypes.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

// for waiting 500μs
static struct timespec req, rem;

static void beat() {
    // set up our timespec structs
    if(req.tv_nsec != BEAT) {
        req.tv_sec = 0;
        req.tv_nsec = BEAT;
    }

    nanosleep(&req, &rem);
}

/** Userland I2C bus read/write callback wrappers */

void pca9685_i2c_bus_read(pca9685_s *h, u8 r) {
    u8 result = h->bus_reader(h, r);
    h->command = "i2c_read";
    h->status = "ok";
    h->address = r;
    h->data = result;
}

void pca9685_i2c_bus_write(pca9685_s *h, u8 r, u8 d) {
    fprintf(stdout, "\nWriting %X to register %X\n", d, r);
    u8 result = h->bus_writer(h, r, d);
    h->command = "i2c_write";
    h->status = result == 0 ? "ok" : "error";
    h->address = r;
    h->data = d;
}

/** Calculations */

u8 channel_to_register_base(u8 c){
    int const in = (int)c;

    int const channel = (in < MIN_CHANNEL) ? MIN_CHANNEL
        : (in > MAX_CHANNEL) ? MAX_CHANNEL
        : in;

    return (u8)(LED_OFFSET + (MULTIPLIER * channel));
}

static u8 led_low(int v) {
    return (u8)(v & 0xFF);
}

static u8 led_high(int v) {
    return (u8)(v >> 0x08);
}

// Calculates prescale value from a PWM output frequency, in Hz
int calculate_prescale_from_frequency(int frequency){
    // Frequency range is 24Hz–1526Hz, or 3–255 (p. 25)
    frequency = (frequency < FREQ_MIN) ? FREQ_MIN
            : (frequency > FREQ_MAX) ? FREQ_MAX
            : frequency;

    double prescale_value = round((OSCILLATOR / (LED_MAX_STEPS * frequency)) - 1);

    return (int)prescale_value;
}

// Calculate delay from percentage
int calculate_delay_time_from_percentage(int d) {
    // Datasheet pp. 17 & 25
    int percent_delay = (d < 0) ? 0
        : (d > 100) ? 100
        : d;

    return (int)round(LED_MAX_STEPS * ((double)percent_delay / 100));
}

// Calculate on time from percentage
int calculate_on_time_from_percentage(int p) {
    // Datasheet pp. 17 & 25
    int percent_on = (p < 0) ? 0
        : (p > 100) ? 100
        : p;

    return (int)round((percent_on * (double)LED_MAX_STEPS) / 100);
}

// calculate off steps from delay and on time
int calculate_off_time_from_delay_and_on_time(int d, int t) {
    int off_time = (d + t) - 1;

    off_time = (off_time <= 0) ? LED_MAX_BITS
        : (off_time > LED_MAX_STEPS) ? (off_time - LED_MAX_STEPS)
        : off_time;

   return off_time;
}

/** Register operations */

void set_led_bytes(pca9685_s *h, int c, int on, int off) {
    u8 channel = (c == ALL) ? (u8)ALL_LED_ON_L : channel_to_register_base((u8)c);

    fprintf(stdout, "\nSetting time ON %d time OFF %d on channel %X\n", on, off, channel);

    pca9685_i2c_bus_write(h, channel, led_low(on));
    pca9685_i2c_bus_write(h, channel + (u8)1, led_high(on));
    pca9685_i2c_bus_write(h, channel + (u8)2, led_low(off));
    pca9685_i2c_bus_write(h, channel + (u8)3, led_high(off));
}

// Sets the value of the PRE_SCALE register with the provided output frequency
void set_pwm_frequency(pca9685_s *h, int frequency) {
    // Datasheet p. 25
    frequency = (frequency < FREQ_MIN) ? FREQ_MIN
        : frequency > FREQ_MAX ? FREQ_MAX
        : frequency;

    const u8 prescale = (u8)calculate_prescale_from_frequency(frequency);

    pca9685_i2c_bus_write(h, MODE1, SLEEP);
    pca9685_i2c_bus_write(h, PRE_SCALE, prescale);
    pca9685_i2c_bus_write(h, MODE1, MODE1_NO_SLEEP);

    beat();
}

void set_pwm_duty_cycle(pca9685_s *h, int c, int d, int p) {
    int delay_time = calculate_delay_time_from_percentage(d);
    int on_time = calculate_on_time_from_percentage(p);
    int off_time = calculate_off_time_from_delay_and_on_time(delay_time, on_time);

    int on_steps = (delay_time <= 0) ? 0 : (delay_time - 1);

    fprintf(stdout, "\nreceived delay %d and on_time %d\n", d, p);
    fprintf(stdout, "\ndelay_time: %d\n", delay_time);
    fprintf(stdout, "\non_time: %d\n", on_time);
    fprintf(stdout, "\non_steps: %d\n", on_time);
    fprintf(stdout, "\noff_time: %d\n", off_time);

    set_led_bytes(h, c, on_steps, off_time);
}

// Reset without having to power cycle (p. 15)
void reset_driver_soft(pca9685_s *h){
    // Set oscillator sleep bit
    pca9685_i2c_bus_write(h, MODE1, SLEEP);

    // default duty cycle and frequency
    set_pwm_duty_cycle(h, ALL, 0, 0);
    set_pwm_frequency(h, 200);

    // check restart bit and reset; saves PWM register contents
    pca9685_i2c_bus_read(h, MODE1);
    u8 value = h->data;
    if(value & RESTART) pca9685_i2c_bus_write(h, MODE1, MODE1_NO_SLEEP);
    beat();
    pca9685_i2c_bus_write(h, MODE1, RESTART);
}

void reset_driver_hard(pca9685_s *h) {
    pca9685_i2c_bus_write(h, MODE1, SLEEP);

    /** Turn off all channels and reset frequency to default */
    set_pwm_duty_cycle(h, ALL, 0, 0);
    set_pwm_frequency(h, 200);

    /** Reset register default values */
    pca9685_i2c_bus_write(h, SUBADR1, SUBADR1_DEFAULTS);
    pca9685_i2c_bus_write(h, SUBADR2, SUBADR2_DEFAULTS);
    pca9685_i2c_bus_write(h, SUBADR3, SUBADR3_DEFAULTS);
    pca9685_i2c_bus_write(h, ALLCALLADR, ALLCALLADR_DEFAULTS);
    pca9685_i2c_bus_write(h, MODE2, MODE2_DEFAULTS);
    pca9685_i2c_bus_write(h, MODE1, MODE1_DEFAULTS);

    beat();
}

